src/flow/src/transform.rs

```diff
@@ -0,0 +1,1221 @@
+use substrait::substrait_proto::proto::rel::RelType;
+use substrait::substrait_proto::proto::{self, plan_rel, Expression, Plan as SubPlan, Rel};
+
+use crate::adapter::error::{
+    DatatypesSnafu, Error, EvalSnafu, InvalidQuerySnafu, NotImplementedSnafu, PlanSnafu,
+    TableNotFoundSnafu,
+};
+use crate::expr::{
+    AggregateExpr, AggregateFunc, BinaryFunc, GlobalId, MapFilterProject, SafeMfpPlan, ScalarExpr,
+    UnaryFunc, UnmaterializableFunc, VariadicFunc,
+};
+use crate::plan::{AccumulablePlan, KeyValPlan, Plan, ReducePlan, TypedPlan};
+use crate::repr::{self, ColumnType, RelationType};
+
+/// a simple macro to generate a not implemented error
+macro_rules! not_impl_err {
+    ($($arg:tt)*)  => {
+        NotImplementedSnafu {
+            reason: format!($($arg)*),
+        }.fail()
+    };
+}
+
+/// generate a plan error
+macro_rules! plan_err {
+    ($($arg:tt)*)  => {
+        PlanSnafu {
+            reason: format!($($arg)*),
+        }.fail()
+    };
+}
+
+pub struct DataflowContext {
+    id_to_name: HashMap<GlobalId, Vec<String>>,
+    name_to_id: HashMap<Vec<String>, GlobalId>,
+    schema: HashMap<GlobalId, RelationType>,
+}
+
+impl DataflowContext {
+    pub fn new() -> Self {
+        Self {
+            id_to_name: HashMap::new(),
+            name_to_id: HashMap::new(),
+            schema: HashMap::new(),
+        }
+    }
+    pub fn register_table(&mut self, id: GlobalId, name: Vec<String>) {
+        self.id_to_name.insert(id, name.clone());
+        self.name_to_id.insert(name, id);
+        todo!("Table provider etc.")
+    }
+
+    /// Retrieves a GlobalId and table schema representing a table previously registered by calling the [register_table] function.
+    ///
+    /// Returns an error if no table has been registered with the provided names
+    pub fn table(&self, name: &Vec<String>) -> Result<(GlobalId, RelationType), Error> {
+        let id = self
+            .name_to_id
+            .get(name)
+            .copied()
+            .with_context(|| TableNotFoundSnafu {
+                name: name.join("."),
+            })?;
+        let schema = self
+            .schema
+            .get(&id)
+            .cloned()
+            .with_context(|| TableNotFoundSnafu {
+                name: name.join("."),
+            })?;
+        Ok((id, schema))
+    }
+}
+
+pub fn from_substrait_plan(ctx: &mut DataflowContext, plan: &SubPlan) -> Result<TypedPlan, Error> {
+    // Register function extension
+    let function_extension = plan
+        .extensions
+        .iter()
+        .map(|e| match &e.mapping_type {
+            Some(ext) => match ext {
+                MappingType::ExtensionFunction(ext_f) => Ok((ext_f.function_anchor, &ext_f.name)),
+                _ => not_impl_err!("Extension type not supported: {ext:?}"),
+            },
+            None => not_impl_err!("Cannot parse empty extension"),
+        })
+        .collect::<Result<HashMap<_, _>, Error>>()?;
```

Comment:
It seems that extensions play a crucial role. Can they be extracted into a function and explained in detail in the doc comment (in a way that is friendly to beginners like me)?